### PR TITLE
VL-124_Composables-ts-migration_Dmytro-Holdobin

### PR DESCRIPTION
**Pull request by Explicit12**
> https://ilevel.atlassian.net/browse/VL-124

### CODE REVIEW

File: src/adatper.locale/vue-i18n.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import type { I18n } from "vue-i18n";
```
Comment (itsJohnnyGrid) [inline src/adatper.locale/vue-i18n.ts:1]: if user will not have an `vue-i18n` package it invokes error.

Please remove this type and use somthing more general. For our library case -- it's ok.

File: src/adatper.locale/vueless.ts [hunk 2]
```diff
@@ -1,11 +1,34 @@
 import { shallowRef, ref } from "vue";
 import { merge } from "lodash-es";
 
-import en from "./locales/en.js";
+import en from "./locales/en.ts";
+
+import type { Ref } from "vue";
+
+export interface LocaleMessages {
+  [key: string]: LocaleMessages | string;
```
Comment (itsJohnnyGrid) [inline src/adatper.locale/vueless.ts:9]: type rely to itself? Is it valid? :)

File: src/adatper.locale/vueless.ts [hunk 3]
```diff
@@ -68,51 +99,67 @@ function createTranslateMessageFunction(current, fallback, messages) {
...
-export function getObjectValueByPath(obj, path, fallback) {
+export function getObjectValueByPath<T, K = unknown>(
+  obj: T,
+  path?: string,
+  fallback?: K,
+): K | undefined {
   if (obj == null || !path || typeof path !== "string") return fallback;
-  if (obj[path] !== undefined) return obj[path];
+
+  if ((obj as Record<string, unknown>)[path] !== undefined) {
```
Comment (itsJohnnyGrid) [inline src/adatper.locale/vueless.ts:127]: maybe its better to move this variable with type into vatiable?
`(obj as Record<string, unknown>)[path]`

File: src/adatper.locale/vueless.ts [hunk 4]
```diff
@@ -68,51 +99,67 @@ function createTranslateMessageFunction(current, fallback, messages) {
...
+    obj = (obj as Record<string | number, unknown>)[path[i]] as T;
   }
 
   if (obj == null) {
     return fallback;
   }
 
-  return obj[path[last]] === undefined ? fallback : obj[path[last]];
+  return (
+    (obj as Record<string | number, unknown>)[path[last]] === undefined
```
Comment (itsJohnnyGrid) [inline src/adatper.locale/vueless.ts:161]: same about variable

File: src/composablesTs/useAutoPosition.ts [hunk 5]
```diff
@@ -1,77 +1,66 @@
...
-export enum Position {
-  Left = "left",
-  Right = "right",
-  Top = "top",
-  Bottom = "bottom",
-  Auto = "auto",
+import type { ComputedRef, MaybeRef, Reactive } from "vue";
+
+export enum POSITION {
+  left = "left",
```
Comment (itsJohnnyGrid) [inline src/composablesTs/useAutoPosition.ts:7]: Enum name should be PascalCase
Enums keys should be UPPER_CASE or PascalCase related to the TS conventions.
I prefer PascalCase.

follow ts name convention for enums: PascalCase and for enum keys PascalCase
https://aykhanhuseyn.medium.com/typescript-naming-conventions-crafting-maintainable-code-7d872234fe17

File: src/composablesTs/useAutoPosition.ts [hunk 6]
```diff
@@ -1,77 +1,66 @@
...
 
-// TODO: Remove after full TS migration, use enum instead.
-export const POSITION = {
-  left: "left",
-  right: "right",
-  top: "top",
-  bottom: "bottom",
-  auto: "auto",
-};
+export type Direction = "left" | "right" | "top" | "bottom" | "auto";
```
Comment (itsJohnnyGrid) [inline src/composablesTs/useAutoPosition.ts:14]: Code duplication.

Create union type form enum instead.

```export type Direction = `${POSITION}`;```

File: src/composablesTs/useBreakpoint.ts [hunk 7]
```diff
@@ -3,32 +3,34 @@ import { isSSR } from "../utilsTs/utilHelper.ts";
...
-
-const BREAKPOINT = {
-  xs: 0,
-  sm: 640,
-  md: 768,
-  lg: 1024,
-  xl: 1280,
-  "2xl": 1536,
-};
+enum BREAKPOINT_NAME {
```
Comment (itsJohnnyGrid) [inline src/composablesTs/useBreakpoint.ts:6]: follow ts name convention for enums: PascalCase and for enum keys PascalCase
https://aykhanhuseyn.medium.com/typescript-naming-conventions-crafting-maintainable-code-7d872234fe17

in this case keys is better to be in UPPER_CASE: `BreakpointName.XS`

File: src/composablesTs/useMutationObserver.ts [hunk 8]
```diff
@@ -1,41 +1,44 @@
-import { onBeforeUnmount, onMounted, toValue, watch } from "vue";
+import { computed, onBeforeUnmount, onMounted, toValue, watch } from "vue";
 import { isSSR } from "../utilsTs/utilHelper.ts";
 
-import type { Ref } from "vue";
+import type { TemplateRefElement } from "src/types.ts";
```
Comment (itsJohnnyGrid) [inline src/composablesTs/useMutationObserver.ts:4]: `src/types.ts` - wrong file path

File: src/composablesTs/useMutationObserver.ts [hunk 9]
```diff
@@ -1,41 +1,44 @@
...
-  callBack: () => void,
+  target: TemplateRefElement,
+  callback: MutationCallback,
   config = { childList: true, attributes: true, characterData: true },
 ) {
   if (isSSR) return;
 
-  const observer = new MutationObserver(callBack);
+  const observer = new MutationObserver(callback);
+  const targetValue = computed(() => toValue(target));
```
Comment (itsJohnnyGrid) [inline src/composablesTs/useMutationObserver.ts:14]: what the reason to get value in a computed?

`toValue(target)` is same that `targetValue.value` but shorter.

File: src/adatper.locale/vue-i18n.ts (hunk 1 above)
Comment (Explicit12) [reply src/adatper.locale/vue-i18n.ts:1]: That's i18n adapter, and it's supposed to be used with i18n only.

File: src/adatper.locale/vueless.ts (hunk 2 above)
Comment (Explicit12) [reply src/adatper.locale/vueless.ts:9]: Yes

File: src/composablesTs/useMutationObserver.ts (hunk 9 above)
Comment (Explicit12) [reply src/composablesTs/useMutationObserver.ts:14]:
```
    if (Array.isArray(toValue(target))) {
      toValue(target).value.forEach((element) => {
        observer.observe(element, config);
      });
    } else {
      observer.observe(toValue(target), config);
    }
```
This code is not valid. toValue is function and it return a result each time, so values in If condition and  If block are not the same and typescript cannot be sure if it's array or not.

File: src/adatper.locale/vue-i18n.ts (hunk 1 above)
Comment (itsJohnnyGrid) [reply src/adatper.locale/vue-i18n.ts:1]: reasonable, ok

File: src/composablesTs/useMutationObserver.ts (hunk 9 above)
Comment (itsJohnnyGrid) [reply src/composablesTs/useMutationObserver.ts:14]: Okay

**Review by itsJohnnyGrid — APPROVED** [review]

**Pull request merged**